ignite/metrics/metric.py

```diff
@@ -142,12 +142,44 @@ def completed(self, engine: Engine, name: str) -> None:
         engine.state.metrics[name] = result
 
     def attach(self, engine: Engine, name: str) -> None:
+        """
+        Attaches current metric to provided engine. On the end of engine's run,
+        `engine.state.metrics` dictionary will contain computed metric's value under provided name.
+
+        .. code-block:: python
+
+            metric = ...
+            metric.attach(engine, "mymetric")
+
+            assert "mymetric" in engine.run(data).metrics
+
+        """
         engine.add_event_handler(Events.EPOCH_COMPLETED, self.completed, name)
         if not engine.has_event_handler(self.started, Events.EPOCH_STARTED):
             engine.add_event_handler(Events.EPOCH_STARTED, self.started)
         if not engine.has_event_handler(self.iteration_completed, Events.ITERATION_COMPLETED):
             engine.add_event_handler(Events.ITERATION_COMPLETED, self.iteration_completed)
 
+    def detach(self, engine: Engine) -> None:
+        """
+        Detaches current metric from the engine and no metric's computation is done during the run.
+        This method in conjunction with :meth:`~ignite.metrics.Metric.attach` can be useful if several
+        metrics need to be computed with different periods. For example, one metric is computed every training epoch
+        and another metric (e.g. more expensive one) is done every n-th training epoch.
+        """
+        if engine.has_event_handler(self.completed, Events.EPOCH_COMPLETED):
```

Comment:
Probably, we need also raise something if metric is not attached but we call detach.

Reply:
I checked and it doesn't work fine. `MetricLambda` requires that inner metrics be attached. In your example, when `cm_metric` is detached, `started` and `iteration_completed` methods are no longer called during iterations. So accumulation for `cm_metric` is frozen (to deprecated state). But reaching event `EPOCH_COMPLETED`, `iou` will call `compute` method of `cm_metric`... It's a bug situation.
A partial fix should be to check during `compute` if metrics are always attached and raise an exception (or warning) if not.

Reply:
Not so simple 
https://github.com/pytorch/ignite/blob/a22a0f5e909ac70d2a1f76a60b6e84b2134f196c/ignite/metrics/metrics_lambda.py#L61-L64
To detach, engine is needed. We don't have the engine.

Reply:
Yes, that's what I thought that potentially we can have problems with `MetricsLambda`...
Maybe we can leave this to user...

Reply:
I commited a fix

Reply:
I commited a version using weakref

Reply:
Thanks! Looks good!